### PR TITLE
Infer shape of transposition operations using the perm attribute

### PR DESCRIPTION
**Pull request by doru1004**
This patch takes into consideration the attribute of the transposition operation when inferring the shape of the output tensor.

### CODE REVIEW

File: src/dialect/onnx/onnx_ops.cpp [hunk 1]
```diff
@@ -407,11 +407,25 @@ void ONNXTransposeOp::inferShapes() {
...
-  SmallVector<int64_t, 2> dims(llvm::reverse(arrayTy.getShape()));
-  getResult().setType(RankedTensorType::get(dims, arrayTy.getElementType()));
+  auto arrayTy = getOperand()->getType().cast<RankedTensorType>();
+  SmallVector<int64_t, 2> dims;
+
+  if (auto permutation = getAttrOfType<ArrayAttr>(
+          ONNXTransposeOp::getPermAttrName())) {
+    // Perform transposition according to perm attribute.
+    for (auto perm : permutation.getValue()) {
+      int32_t index = perm.cast<IntegerAttr>().getInt();
```
Comment (tjingrant) [inline src/dialect/onnx/onnx_ops.cpp]: Integer in onnx attribute are int64 (https://github.com/onnx/onnx/blob/a32db1c5233c0b7c124c1529ccd026f1d3f17040/onnx/onnx-ml.proto#L147).

File: src/dialect/onnx/onnx_ops.cpp [hunk 2]
```diff
@@ -407,11 +407,25 @@ void ONNXTransposeOp::inferShapes() {
...
+      if (index < 0)
```
Comment (tjingrant) [inline src/dialect/onnx/onnx_ops.cpp]: I feel like attribute validation can potentially be implemented as op verifier.
But that'd be another patch. What're your thoughts?

File: src/dialect/onnx/onnx_ops.cpp [hunk 3]
```diff
@@ -407,11 +407,25 @@ void ONNXTransposeOp::inferShapes() {
...
+        emitError("Cannot tranpose when permutation contains negative index.");
+      dims.emplace_back(arrayTy.getShape()[index]);
+    }
+  } else {
+    // Default
+    for (auto shape : llvm::reverse(arrayTy.getShape()))
```
Comment (tjingrant) [inline src/dialect/onnx/onnx_ops.cpp]: shape -> dim_size?
Or something more accurate and concise.

File: src/dialect/onnx/onnx_ops.cpp (hunk 2 above)
Comment (doru1004) [reply src/dialect/onnx/onnx_ops.cpp]: I agree, actually I've started doing that for the Conv case.
